lotus-soup/testkit/node_time.go

```diff
@@ -0,0 +1,103 @@
+package testkit
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/raulk/clock"
+	"github.com/testground/sdk-go/sync"
+)
+
+// ClockSyncTopic is the topic where clock synchronization events are exchanged.
+var ClockSyncTopic = sync.NewTopic("clock_sync", &ClockSyncMsg{})
+
+// ClockSyncMsg is a clock synchronization message.
+type ClockSyncMsg struct {
+	ID    string // ${groupname}_${groupseq}
+	Epoch abi.ChainEpoch
+}
+
+// SynchronizeClock synchronizes this instance's clock with the clocks of all
+// miner and client nodes participating in this test scenario.
+//
+// This function takes a channel (localEpochStartCh) where we expect to receive
+// local signals to advance the clock to the next epoch. Such signals should
+// correspond to some local event that implies readiness to advance, such as
+// "mining round completed" events from synchronized mining.
+//
+// Upon receiving a local advance signal, it forwards the clock by build.BlockDelaySecs,
+// thus progressing one chain epoch. It then signals this fact to all other
+// nodes participating in the test via the synchronization service.
+//
+// Once all miners and clients have advanced, this function releases the new
+// abi.ChainEpoch on the `globalEpochStartCh`, thus signalling that all nodes
+// are ready to commence that chain epoch.
+//
+// Essentially this behaviour can be assimilated to a distributed sempaphore,
+// where all miners and clients wait until they have locally advanced their
+// clocks. Once they all signal that fact via the relevant sync topic, the
+// semaphore fires and allows them all to proceed at once.
+//
+// All of this happens in a background goroutine.
+func (n *LotusNode) SynchronizeClock(ctx context.Context, genesisTime time.Time, localEpochStartCh <-chan abi.ChainEpoch, globalEpochStartCh chan<- abi.ChainEpoch) {
+	// replace the clock, setting it to the genesis timestamp + 1 epoch
+	// we are now ready to mine (and receive) the first block.
+	n.MockClock = clock.NewMock()
+	build.Clock = n.MockClock
+
+	// start at genesis + 100ms.
+	n.MockClock.Set(genesisTime.Add(100 * time.Millisecond))
```

Comment:
This still triggers (after last set of patches) the `mined block in the past` and related. In total we are advancing `BlockDelaySecs` plus this slight drift, putting us outside of the mining epoch window: the miner only advances `BlockDelaySecs` from its tipset and sees that it lost the mining window by 100ms. Ideally this should be a subtraction but in my local tests this breaks other assumptions (understandably because I'm drifting before genesis time). Maybe we should jump a smaller amount than `BlockDelaySecs` the first time (to remain inside the mining window when adding the full `BlockDelaySecs` amount in subsequent rounds).

Reply:
I think the simplest solution here would be to advance the clock _after_ the mining round and not before as it seems to be done now. (I think the change would translate to something like moving the clock advance logic from the local to the global case in `SynchronizeClock`.)

Reply:
That would make sense. I did try different variants and in some cases I ended up hitting the "got a block from the future" error. I think we do need to get this _absolutely right_, so I'm willing to invest time in nailing the state machine right now so that we're all happy and we can build on a solid base.

Reply:
I think advancing the clock _just before_ the mining round won't work because of races between miners. Miners who advance their clock sooner will end up propagating what appear to be "blocks from the future" to miners who still haven't advanced their clocks. That's why I opted for this approach, as we have the assurance that all miners are in the right epoch when mining commences in concert.

After removing the drift, and testing atop [this patch](https://github.com/filecoin-project/lotus/pull/2362/files/55d88440eb5cbbd31950bdf664b50b5456372ede#diff-7fc2e947b54d13e96c3477dc4e6e6c06), we no longer have the spurious logs.

The only one we're left with is the occasional:

```
WARN	sub	sub/incoming.go:81	Received block with large delay 25 from miner t01001	{"req_id": "a305c648"}
```

Which I think has to do with gossipsub propagation, and not worth investigating right now IMO.